File: app/javascript/game/models/player.ts
export default class Player {
  public turnOver: boolean;

  readonly currentPlayer: boolean;
  readonly game_id: string;
  readonly host: boolean;
  readonly number: number;
  readonly role: string;
  readonly user_id: string;
  readonly username: string;
  readonly civilian_count?: number;
  readonly military_count?: number;
  readonly growth?: number;
  readonly supply?: number;
  readonly id?: string;
  readonly raw_user_id?: string;

  constructor(rawPlayerData: RawPlayerData) {
    this.number = rawPlayerData.number;
    this.currentPlayer = rawPlayerData.current_player;
    this.turnOver = rawPlayerData.turn_over;
    this.game_id = rawPlayerData.game_id;
    this.host = rawPlayerData.host;
    this.number = rawPlayerData.number;
    this.role = rawPlayerData.role;
    this.user_id = rawPlayerData.user_id;
    this.username = rawPlayerData.username;
    this.civilian_count = rawPlayerData.civilian_count;
    this.military_count = rawPlayerData.military_count;
    this.growth = rawPlayerData.growth;
    this.supply = rawPlayerData.supply;
    this.id = rawPlayerData.id;
    this.raw_user_id = rawPlayerData.raw_user_id;
  };
}

File: app/models/player/player.rb
class Player
  include CivCloneMongoModel
  include Mongoid::Document
  include Player::Lobby
  include Player::Logic

  embedded_in :game
  
  direct_children :vision_squares

  embeds_many :diplomatic_stances
  embeds_many :vision_squares, class_name: "Square::Vision"
  
  field :number,         type: Integer
  field :turn_over,      type: Boolean, default: false
  field :growth,         type: Integer, default: 1
  field :supply,         type: Integer, default: 0
  field :military_count, type: Integer, default: 0
  field :civilian_count, type: Integer, default: 0

  field :role, type: String, default: "player"
  field :host, type: Boolean, default: false
  field :raw_user_id, type: String
  field :username, type: String
  
  validates :role, inclusion: { in: ["player", "dead_player", "observer"] }
end